docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx

```diff
@@ -0,0 +1,1477 @@
+  properties: (SdkEndpointParameter | SdkCredentialParameter | SdkMethodParameter)[];
+}
+```
+
+### Example
+
+```ts
+const sdkInitializationType = {
+  name: "WidgetManagerOptions",
+  kind: "model",
+  properties: [
+    {
+      kind: "endpoint",
+      urlEncode: false,
+      onClient: true,
+    },
+    {
+      kind: "credential",
+      type: {
+        kind: "credential",
+        scheme: {
+          kind: "apiKey",
+          in: "header",
+          name: "api-key"
+        }
+      },
+      onClient: true,
+    },
+  ],
+  crossLanguageDefinitionId: "Contoso.WidgetManager.WidgetManagerOptions",
+  apiVersions: [],
+  usage: UsageFlags.Input,
+  access: "public",
+  nullable: false,
+  isFormDataType: false,
+  isError: false,
+}
+
+### Usage
+
+The usage of this property is more language-dependent.
+
+Some emitters will create a model and have the client accept the model as options to initialize your client. Others will flatten out the parameters and accept them individually as parameter input.
+
+## SdkMethod
+
+An `SdkMethod` is any of the types of methods that can be a method on a client.
+
+```ts
+export type SdkMethod<
+  TServiceOperation extends PythonSdkServiceOperation
+> =
+  SdkServiceMethod<TServiceOperation> |
+  SdkClientAccessor<TServiceOperation>;
+```
+
+They each extend from the shared `SdkMethodBase`
+
+```ts
+interface SdkMethodBase<
+  TServiceOperation extends SdkServiceOperation
+> {
+  __raw?: Operation;
+  name: string;
+  access: AccessFlags | undefined;
+  parameters: SdkParameter[];
+  apiVersions: string[];
+  description?: string;
+  details?: string;
+  overloads?: SdkMethod<TServiceOperation>[];
+  overloading?: SdkMethod<TServiceOperation>;
+}
+```
+
+Below we go into each method type
+
+### SdkClientAccessor
+
+A `clientaccessor` method is simply a method on a client that returns another client. The returned client can be instantiable or uninstantiable. If the returned client is instantiable, most likely your `clientaccessor` will be part of the public api surface, so users can instantiate the subclient using your client method. If it is not instantiable, it is up to you how to expose the subclient on your current client for users to access.
+
+#### Interface
+
+```ts
+interface SdkClientAccessor<
+  TServiceOperation extends SdkServiceOperation
+>
+  extends SdkMethodBase<TServiceOperation> {
+    kind: "clientaccessor";
+    response: SdkClientType<TServiceOperation>;
+  }
+```
+
+#### Example
+
+```ts
+const sdkClientAccessor = {
+  name: "getWidgets",
+  kind: "clientaccessor",
+  response: {
+    name: "Widgets",
+    kind: "client",
+    initialization: undefined,
+    ...
+  }
+}
+```
+
+### SdkServiceMethod
+
+An `SdkServiceMethod` is any service method on a client that calls the service.
+
+The actual service call is a separate property on the method, `.operation`. This way, our service methods are able to abstract away the protocol used to call the service (i.e. `http` or `gRPC`)
```

Comment:
Currently for Python adoption, we do not relay on `SdkServiceMethod` info. Do we have any plan to leverage it?

Reply:
we should leverage `SdkServiceMethod` for creating the method signature, eventually we should fully move over to this structure bc it gives us much more flexibility. Our hands are kind of tied right now with trying to maintain the old Python SdkType structure